File: Java/Dijkstra/Dijkstra.java
/*
 * Implementation of Dijkstra's algorithm for finding
 * the shortest path. Also this class is responsible
 * for providing interface to GUI programs.

 * Implementation details:
    1. Intilialize the source vertex as solved.
       Lable it with '0', since distance to itself is 0.
    2. Identify all unsolved vertices connected to any solved vertex.
    3. For each edge connecting a solved vertex and unsolved vertex,
       calculate the candidate distance.
       candidate distance = distance to the solved vertex + length of edge.
    4. Choose the smallest candidate distance.
       Change it to solved and label it with candidate distance.
    5. Add the edge to the edge set.
    6. Repeat step 2-5 until we have reached the destination vertex.
    
 */

import java.util.Iterator;
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.Comparator;
import java.util.Set;

public class Dijkstra {

    private Graph graph = null;
    private Vertex start = null;
    private Vertex end = null;

    // We use a priority queue to keep track of unsolved vertices.
    private PriorityQueue<Vertex> pqueue;
    // Used to store all solved vertice.
    private ArrayList<Vertex> solved;
    // Used to keep track of the shortest path.
    private Map<Vertex, Vertex> prev;

    public Dijkstra(Graph graph, Vertex start, Vertex end) {
	this.graph = graph;
	this.start = start;
	this.end = end;

	Comparator<Vertex> comparator = new DistComparator();
	pqueue = new PriorityQueue<Vertex>(graph.getVertexNum(), comparator);
	solved = new ArrayList<Vertex>();
	prev = new HashMap<Vertex, Vertex>();

	initPQueue();
    }

    /*
     * Initiliaze the priority queue 'pq'.
     */
    private void initPQueue() {
	Set<Vertex> vset = graph.getVertice();
	Iterator it = vset.iterator();

	while (it.hasNext()) {
	    Vertex v = (Vertex) it.next();
	    if (v.cmp(start)) {
		// Start vertex has distance of zero to itself.
		v.setDistToStart(0);
	    }
	    pqueue.add(v);
	}

	/*
	while (pqueue.size() != 0) {
	    System.out.println("Priority Queue:");
	    Vertex u = pqueue.remove();
	    System.out.println(u.getLabel() + " : " + u.distToStart());
	}
	*/
    }

    public void traverse() {

	while (!this.pqueue.isEmpty()) {
	    // Pop up the vertex with smallest distance
	    // to the start vertex.
	    Vertex u = this.pqueue.remove();
	    int dist = u.distToStart();

	    if (u.cmp(this.end)) {
		// If we have popped up the end vertex, we are done.
		System.out.println("Shortest path found!");
		System.out.println("Shortest distance is " + dist);
	    }

	    ArrayList<Vertex> neighbors = this.graph.findNeighbors(u);
	    for (int i = 0; i < neighbors.size(); i++) {
		Vertex v = neighbors.get(i);
		if (this.solved.contains(v)) {
		    // Ignore the vertex that has been solved.
		    continue;
		}

		int alt_dist = dist + this.graph.distBetween(u, v);
		if (alt_dist < v.distToStart()) {
		    this.pqueue.remove(v);
		    v.setDistToStart(alt_dist);
		    this.pqueue.add(v);
		    // Vertex 'u' is the previous vertex of 'v'.
		    // We need to keep track of it so that we can
		    // retrieve the final shortest path.
		    this.prev.put(v, u);
		}
	    }

	    // Updated arrayList of solved vertice.
	    this.solved.add(u);
	}
    }

    public void printPath() {
	ArrayList<Vertex> path = new ArrayList<Vertex>();
	Vertex v = this.graph.getVertex(this.end);

	path.add(v);
	while (this.prev.containsKey(v)) {
	    v = this.prev.get(v);
	    path.add(v);
	}

	for (int i = path.size()-1; i >= 0; i--) {
	    path.get(i).printSelf();
	}
    }
}

File: Python/Dijkstra/Dijkstra.py
#!/usr/bin/env python

"""
Implementation of Dijkstra's algorithm.

Dijkstra's algorithm operates by maintaining a subset of
vertices 'S' of the whole set of vertex 'V',
for which we know the true distance
d[v] = (shortest distance between source vertex 's' and 'v').

Intially 'S' = empty set. And we set d[s] = 0 and d[v] = -1 for
all others vertices. One by one we select vertices from 'V' to 'S'.

The set 'S' is implemented by an array of vertex colors.
Intially all vertices are 'white', and we set color[v] = 'black'
to indicate that 'v' belongs to 'S'.

Dijkstra's algorithm is a greedy algorithm(in order to make sure
we can always find a shortest path or shortest path tree).
For each vertex 'u' belongs to 'V'\'S'(means 'V'-'S', that is,
the subset of vertices we haven't processed), we have computed a
distance estimate d[u]. The next vertex processed is always a
vertex for which d[u] is minimum(For the very first time, 'u' is
source vertex 's'), that is, we take the unprocessed vertex that
is closest(by our estimate) to 's'.

"""

import sys
from operator import itemgetter
from Graph import *
from heapq import *

class Dijkstra(object):
    """
    Steps:
    1. Intilialize the source vertex as solved.
       Lable it with '0', since distance to itself is 0.
    2. Identify all unsolved vertices connected to any solved vertex.
    3. For each edge connecting a solved vertex and unsolved vertex,
       calculate the candidate distance.
       candidate distance = distance to the solved vertex + length of edge.
    4. Choose the smallest candidate distance.
       Change it to solved and label it with candidate distance.
    5. Add the edge to the edge set.
    6. Repeat step 2-5 until we have reached the destination vertex.
    
    """
    def __init__(self, graph, src_v, dest_v):
        """
        'graph' is a Graph object defined in Graph.py.
        'src_v' is the source vertex, take vertex 'O' as default.
        'dest_v' is the destination vertex, take vertex 'T' as default.
        """
        ## We use a dict to map each vertex 'v' to its distance 'd'
        ## where each v belongs to 'V'\'S'(subset of vertices not processed)
        self.v2d = {}

        self.graph = graph
        self.src_v = src_v
        self.dest_v = dest_v

        self.solved_vs = []

        ## keep track of the shortest path.
        self.previous = {}
        
        self.init_v2d()

    def init_v2d(self):
        """
        Initialize the priority queue. Set distance for source vertex
        as zero, and 'sys.maxint' for all other vertices.
        """
        for v in self.graph.keys():
            if v == self.src_v:
                self.v2d[v] = 0
            else:
                self.v2d[v] = sys.maxint

    def update_v2d(self, v, d):
        for key in self.v2d.keys():
            if key == v:
                self.v2d[v] = d

    def traverse(self):
        while self.v2d:
            ## pop up a vertex 'u' with shortest distance to source vertex.
            u, dist_u = sorted(self.v2d.iteritems(), key=itemgetter(1))[0]
            
            if dist_u == sys.maxint:
                print 'No vertex has the true distance.'
                break

            if u == self.dest_v:
                print '\nShortes path from %s to %s has been found' % (self.src_v, self.dest_v)
                print 'Shortest distance is', dist_u
                print
                self.shortest_path()
                break

            ## Pop up 'u'.
            for key in self.v2d.keys():
                if key == u:
                    self.v2d.pop(key)

            neighbor_vs = self.graph.find_neighbors(u)
            for v in neighbor_vs:
                if v in self.solved_vs:
                    continue
                
                alt_dist = dist_u + self.graph.dist_between(u, v)
                if alt_dist < self.v2d[v]:
                    self.update_v2d(v, alt_dist)
                    self.previous[v] = u

            self.solved_vs.append(u)

    def shortest_path(self):
        path = []
        v = self.graph.get_vertex(self.dest_v)

        path.append(v)
        while self.previous.has_key(v):
            path.append(self.previous[v])
            v = self.previous[v]

        path.reverse()
        print 'The shortest path:'
        print ' ==> '.join(map(str, path))
        print

def main():
    vs, es = load_graph()
    g = Graph(vs, es)

##    dk = Dijkstra(g, Vertex('O'), Vertex('T'))
    dk = Dijkstra(g, Vertex('E'), Vertex('A'))
    dk.traverse()

if __name__ == '__main__':
    main()
